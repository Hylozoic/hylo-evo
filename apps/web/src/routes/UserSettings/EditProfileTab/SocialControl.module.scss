// SocialControl
.control {
  composes: control from 'components/SettingsControl/SettingsControl.module.scss';
}

.linkedIcon {
  color: rgba(13, 195, 159, 1.000);
}

.social-control-label {
  composes: control-input from 'components/SettingsControl/SettingsControl.module.scss';
  display: flex;
  align-items: center;
}

.link-button {
  composes: text-button from 'css/typography.scss';
  margin-left: auto;
}

@media screen and (max-width: 425px) {
  .save-button {
    font-size: 12px;
  }
}

@media screen and (max-width: 320px) {
  .saveChanges {
    span {
      padding: 0;
      font-size: 12px;
      line-height: 12px;
    }
  }
  .save-button {
    font-size: 12px;
  }
}
