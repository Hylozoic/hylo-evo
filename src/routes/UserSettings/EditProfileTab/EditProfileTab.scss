// EditProfile
.name {
  height: 60px;
  line-height: 60px;
  text-align: center;
  border-radius: 4px;
  margin-bottom: 25px;
  composes: hdr-display from 'css/typography.scss';
  background-color: $color-catskill-white;
  border: none;
  width: 100%;
}

.name-validation {
  color: red;
  position: absolute;
  top: 30px;
}

.banner {
  width: 100%;
  height: 165px;
  border-radius: 4px;
  background-position: center;
  background-size: cover;
  padding-top: 5px;
  padding-left: 5px;
}

.change-banner-button {
  /* empty */
}

.change-avatar-button {
  margin: 0 auto;
}

.avatar {
  display: flex;
  align-items: flex-end;
  width: 140px;
  height: 140px;
  border-radius: 50%;
  background-position: center;
  background-size: cover;
  margin: -70px auto $space-8x auto;
  cursor: pointer;

  .uploadIcon {
    margin: 0 auto;
  }
}

.uploadIcon {
  cursor: pointer;
  opacity: 0.5;
  font-size: 30px;
  color: $color-white;
  text-shadow: 1px 1px 2px rgba(0, 0, 0, 0.5);
}

.social-label {
  composes: control-label from 'components/SettingsControl/SettingsControl.scss';
  margin-top: $space-9x;
}

.button-row {
  display: flex;
  width: 100%;
  margin-top: $space-9x;
  margin-bottom: 100px;
}

.saveChanges {
  position: fixed;
  bottom: 10px;
  padding: 10px;
  background: rgba(255,255,255,1);
  box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.15);
  border-radius: 5px;
  display: flex;
  align-items: center;
  width: inherit;
  margin-top: $space-9x;

  span {
    display: inline-block;
    color: rgba(35, 64, 91, .5);
    margin-right: 30px;
    font-size: 14px;
  }

  span.settingChanged {
    color: rgba(35, 64, 91, 1);
  }
}

.save-button {
  margin: 0 auto;
}
