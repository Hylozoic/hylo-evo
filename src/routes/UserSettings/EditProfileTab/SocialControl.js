import React, { Component } from 'react'
import { withTranslation } from 'react-i18next'
import PropTypes from 'prop-types'
import cx from 'classnames'
import Icon from 'components/Icon'
import './SocialControl.scss'

const { func, string } = PropTypes

class SocialControl extends Component {
  static propTypes = {
    label: string,
    provider: string,
    value: string,
    updateSettingDirectly: func,
    handleUnlinkAccount: func,
    onLink: func,
    fetchLocation: func
  }

  windowPrompt (network, urlPattern) {
    const promptText = this.props.t(`Please enter the full url for your {{network}} page.`, { network })
    const invalidUrlText = this.props.t(`Invalid url. Please enter the full url for your {{network}} page.`, { network })

    let url = window.prompt(promptText)

    if (url) {
      while (!url.match(urlPattern)) {
        url = window.prompt(invalidUrlText)
      }
    }
    return url
  }

  handleLinkClick () {
    const { provider, updateSettingDirectly } = this.props

    switch (provider) {
      case 'twitter': {
        const twitterHandle = window.prompt(this.props.t('Please enter your twitter name.'))
        if (twitterHandle) {
          updateSettingDirectly()(twitterHandle)
        }
        break
      }
      case 'linkedin': {
        const network = 'LinkedIn'
        const urlPattern = /^(http(s)?:\/\/)?([\w]+\.)?linkedin\.com/
        const url = this.windowPrompt(network, urlPattern)
        if (url) {
          updateSettingDirectly()(url)
        }
        break
      }
      case 'facebook': {
        const network = 'Facebook'
        const urlPattern = /^(http(s)?:\/\/)?([\w]+\.)?facebook\.com/
        const url = this.windowPrompt(network, urlPattern)
        if (url) {
          updateSettingDirectly()(url)
        }
        break
      }
    }
  }

  handleUnlinkClick () {
    const { handleUnlinkAccount, updateSettingDirectly } = this.props

    handleUnlinkAccount()
    updateSettingDirectly()(null)
  }

  render () {
    const { label, value = '', provider, onLink, t } = this.props
    const linked = !!value

    const linkButton =
      <span
        styleName='link-button'
        onClick={linked ? () => this.handleUnlinkClick() : () => this.handleLinkClick()}>
        {linked ? t('Unlink') : t('Link')}

      </span>

    const connectFacebookButton =
      <span
        styleName='link-button'
        onClick={linked ? () => this.handleUnlinkClick() : () => onLink()}
        className='ml-auto'
      >
        {linked ? t('Disconnect') : t('Connect')}
      </span>

    return (
      <div styleName='control'>
        <div styleName={cx('social-control-label')}>
          {linked ? <Icon name='Complete' styleName='linkedIcon' /> : ''}
          {label}
          {provider === 'facebook' && connectFacebookButton}
          {linkButton}
        </div>
      </div>
    )
  }
}

export default withTranslation()(SocialControl)
