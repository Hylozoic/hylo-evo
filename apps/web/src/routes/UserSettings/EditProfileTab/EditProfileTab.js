import React, { Component } from 'react'
import { withTranslation } from 'react-i18next'
import { get } from 'lodash/fp'
import PropTypes from 'prop-types'
import { Helmet } from 'react-helmet'
import cx from 'classnames'
import SettingsControl from 'components/SettingsControl'
import SkillsSection from 'components/SkillsSection'
import SkillsToLearnSection from 'components/SkillsToLearnSection'
import Button from 'components/Button'
import Icon from 'components/Icon'
import UploadAttachmentButton from 'components/UploadAttachmentButton'
import Loading from 'components/Loading'
import { bgImageStyle } from 'util/index'
import { DEFAULT_BANNER } from 'store/models/Me'
import classes from './EditProfileTab.module.scss'
import { ensureLocationIdIfCoordinate } from 'components/LocationInput/LocationInput.store'
import SocialControl from './SocialControl'

const { object, func } = PropTypes

export const validateName = name => name && name.match(/\S/gm)

class EditProfileTab extends Component {
  static propTypes = {
    currentUser: object,
    updateUserSettings: func
  }

  constructor (props) {
    super(props)
    this.state = { edits: {}, changed: false }
  }

  componentDidMount () {
    this.setEditState()
  }

  componentDidUpdate (prevProps, prevState) {
    if (prevProps.fetchPending && !this.props.fetchPending) {
      this.setEditState()
    }
  }

  setEditState () {
    const { currentUser } = this.props
    if (!currentUser) return

    const {
      name, avatarUrl, bannerUrl, tagline, bio,
      contactEmail, contactPhone, locationObject, location,
      url, facebookUrl, twitterName, linkedinUrl
    } = currentUser

    this.setState({
      edits: {
        name: name || '',
        avatarUrl: avatarUrl || '',
        bannerUrl: bannerUrl || DEFAULT_BANNER,
        tagline: tagline || '',
        bio: bio || '',
        contactPhone: contactPhone || '',
        contactEmail: contactEmail || '',
        location: location || '',
        locationId: get('id', locationObject) || null,
        url: url || '',
        facebookUrl,
        twitterName,
        linkedinUrl
      }
    })
  }

  updateSetting = (key, setChanged = true) => async event => {
    const { fetchLocation, t } = this.props
    const { edits, changed } = this.state
    setChanged && this.props.setConfirm(t('You have unsaved changes, are you sure you want to leave?'))

    if (key === 'location') {
      edits['location'] = event.target.value.fullText
      edits['locationId'] = await ensureLocationIdIfCoordinate({ fetchLocation, location: edits.location, locationId: event.target.value.id })
    } else {
      edits[key] = event.target.value
    }

    this.setState({
      changed: setChanged ? true : changed,
      edits
    })
  }

  updateSettingDirectly = (key, changed) => value =>
    this.updateSetting(key, changed)({ target: { value } })

  save = () => {
    this.setState({ changed: false })
    this.props.setConfirm(false)
    this.props.updateUserSettings(this.state.edits)
  }

  render () {
    const {
      fetchPending,
      currentUser,
      unlinkAccount,
      t
    } = this.props
    if (fetchPending || !currentUser) return <Loading />

    const { edits, changed } = this.state
    const {
      name, avatarUrl, bannerUrl, tagline, bio,
      contactEmail, contactPhone, location, url,
      facebookUrl, twitterName, linkedinUrl
    } = edits

    const locationObject = currentUser.locationObject

    return (
      <div>
        <Helmet>
          <title>{t('Your Settings')} | Hylo</title>
        </Helmet>
        <label className={classes.label}>{t('Your Name')}</label>
        {!validateName(name) && <div className={classes.nameValidation}>{t('Name must not be blank')}</div>}
        <input type='text' className={classes.name} onChange={this.updateSetting('name')} value={name || ''} />
        <label className={classes.label}>{t('Banner and Avatar Images')}</label>
        <UploadAttachmentButton
          type='userBanner'
          id={currentUser.id}
          onSuccess={({ url }) => this.updateSettingDirectly('bannerUrl')(url)}
          className={classes.changeBanner}
        >
          <div style={bgImageStyle(bannerUrl)} className={classes.bannerImage}><Icon name='AddImage' className={classes.uploadIcon} /></div>
        </UploadAttachmentButton>
        <UploadAttachmentButton
          type='userAvatar'
          id={currentUser.id}
          onSuccess={({ url }) => this.updateSettingDirectly('avatarUrl')(url)}
          className={classes.changeAvatar}
        >
          <div style={bgImageStyle(avatarUrl)} className={classes.avatarImage}><Icon name='AddImage' className={classes.uploadIcon} /></div>
        </UploadAttachmentButton>
        <SettingsControl label={t('Tagline')} onChange={this.updateSetting('tagline')} value={tagline} maxLength={60} />
        <SettingsControl label={t('About Me')} onChange={this.updateSetting('bio')} value={bio} type='textarea' />
        <SettingsControl
          label={t('Location')}
          onChange={this.updateSettingDirectly('location', true)}
          location={location}
          locationObject={locationObject}
          type='location'
        />
        <SettingsControl label={t('Website')} onChange={this.updateSetting('url')} value={url} />
        <SettingsControl label={t('My Skills & Interests')} renderControl={() =>
          <SkillsSection personId={currentUser.id} />} />
        <SettingsControl label={t('What I\'m learning')} renderControl={() =>
          <SkillsToLearnSection personId={currentUser.id} />} />
        <SettingsControl label={t('Contact Email')} onChange={this.updateSetting('contactEmail')} value={contactEmail} />
        <SettingsControl label={t('Contact Phone')} onChange={this.updateSetting('contactPhone')} value={contactPhone} />
        <label className={classes.socialLabel}>{t('Social Accounts')}</label>
        <SocialControl
          label='Facebook'
          provider='facebook'
          value={facebookUrl}
          updateSettingDirectly={() => this.updateSettingDirectly('facebookUrl')}
          handleUnlinkAccount={() => unlinkAccount('facebook')}
        />
        <SocialControl
          label='Twitter'
          provider='twitter'
          value={twitterName}
          updateSettingDirectly={() => this.updateSettingDirectly('twitterName')}
          handleUnlinkAccount={() => unlinkAccount('twitter')}
        />
        <SocialControl
          label='LinkedIn'
          provider='linkedin'
          value={linkedinUrl}
          updateSettingDirectly={() => this.updateSettingDirectly('linkedinUrl')}
          handleUnlinkAccount={() => unlinkAccount('linkedin')}
        />
        <div style={{ height: '80px' }} />
        <div className={classes.saveChanges}>
          <span className={cx({ [classes.settingChanged]: changed })}>{changed ? t('Changes not saved') : t('Current settings up to date')}</span>
          <Button label={t('Save Changes')} color={changed && validateName(name) ? 'green' : 'gray'} onClick={changed && validateName(name) ? this.save : null} className={classes.saveButton} />
        </div>
      </div>
    )
  }
}

export default withTranslation()(EditProfileTab)
