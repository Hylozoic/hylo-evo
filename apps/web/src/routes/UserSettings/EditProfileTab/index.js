import EditProfileTab from './EditProfileTab'

export default EditProfileTab
