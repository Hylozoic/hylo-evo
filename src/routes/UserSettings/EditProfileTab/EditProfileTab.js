import React, { Component } from 'react'
import { withTranslation } from 'react-i18next'
import { get } from 'lodash/fp'
import PropTypes from 'prop-types'
import { Helmet } from 'react-helmet'
import SettingsControl from 'components/SettingsControl'
import SkillsSection from 'components/SkillsSection'
import SkillsToLearnSection from 'components/SkillsToLearnSection'
import Button from 'components/Button'
import Icon from 'components/Icon'
import UploadAttachmentButton from 'components/UploadAttachmentButton'
import Loading from 'components/Loading'
import { bgImageStyle } from 'util/index'
import { DEFAULT_BANNER } from 'store/models/Me'
import './EditProfileTab.scss'
import { ensureLocationIdIfCoordinate } from 'components/LocationInput/LocationInput.store'
import SocialControl from './SocialControl'

const { object, func } = PropTypes

export const validateName = name => name && name.match(/\S/gm)

class EditProfileTab extends Component {
  static propTypes = {
    currentUser: object,
    updateUserSettings: func,
    loginWithService: func
  }

  constructor (props) {
    super(props)
    this.state = { edits: {}, changed: false }
  }

  componentDidMount () {
    this.setEditState()
  }

  componentDidUpdate (prevProps, prevState) {
    if (prevProps.fetchPending && !this.props.fetchPending) {
      this.setEditState()
    }
  }

  setEditState () {
    const { currentUser } = this.props
    if (!currentUser) return

    const {
      name, avatarUrl, bannerUrl, tagline, bio,
      contactEmail, contactPhone, locationObject, location,
      url, facebookUrl, twitterName, linkedinUrl
    } = currentUser

    this.setState({
      edits: {
        name: name || '',
        avatarUrl: avatarUrl || '',
        bannerUrl: bannerUrl || DEFAULT_BANNER,
        tagline: tagline || '',
        bio: bio || '',
        contactPhone: contactPhone || '',
        contactEmail: contactEmail || '',
        location: location || '',
        locationId: get('id', locationObject) || null,
        url: url || '',
        facebookUrl,
        twitterName,
        linkedinUrl
      }
    })
  }

  updateSetting = (key, setChanged = true) => async event => {
    const { fetchLocation, t } = this.props
    const { edits, changed } = this.state
    setChanged && this.props.setConfirm(this.props.t('You have unsaved changes, are you sure you want to leave?'))

    if (key === 'location') {
      edits['location'] = event.target.value.fullText
      edits['locationId'] = await ensureLocationIdIfCoordinate({ fetchLocation, location: edits.location, locationId: event.target.value.id })
    } else {
      edits[key] = event.target.value
    }

    this.setState({
      changed: setChanged ? true : changed,
      edits
    })
  }

  updateSettingDirectly = (key, changed) => value =>
    this.updateSetting(key, changed)({ target: { value } })

  save = () => {
    this.setState({ changed: false })
    this.props.setConfirm(false)
    this.props.updateUserSettings(this.state.edits)
  }

  render () {
    const {
      fetchPending,
      currentUser,
      loginWithService,
      unlinkAccount,
      t
    } = this.props
    if (fetchPending || !currentUser) return <Loading />

    const { edits, changed } = this.state
    const {
      name, avatarUrl, bannerUrl, tagline, bio,
      contactEmail, contactPhone, location, url,
      facebookUrl, twitterName, linkedinUrl
    } = edits

    const locationObject = currentUser.locationObject

    return (
      <div>
        <Helmet>
          <title>{t('Your Settings')} | Hylo</title>
        </Helmet>
        <label styleName='label'>{t('Your Name')}</label>
        {!validateName(name) && <div styleName='name-validation'>{t('Name must not be blank')}</div>}
        <input type='text' styleName='name' onChange={this.updateSetting('name')} value={name || ''} />
        <label styleName='label'>{t('Banner and Avatar Images')}</label>
        <UploadAttachmentButton
          type='userBanner'
          id={currentUser.id}
          onSuccess={({ url }) => this.updateSettingDirectly('bannerUrl')(url)}
          styleName='change-banner'
        >
          <div style={bgImageStyle(bannerUrl)} styleName='banner-image'><Icon name='AddImage' styleName='uploadIcon' /></div>
        </UploadAttachmentButton>
        <UploadAttachmentButton
          type='userAvatar'
          id={currentUser.id}
          onSuccess={({ url }) => this.updateSettingDirectly('avatarUrl')(url)}
          styleName='change-avatar'
        >
          <div style={bgImageStyle(avatarUrl)} styleName='avatar-image'><Icon name='AddImage' styleName='uploadIcon' /></div>
        </UploadAttachmentButton>
        <SettingsControl label={t('Tagline')} onChange={this.updateSetting('tagline')} value={tagline} maxLength={60} />
        <SettingsControl label={t('About Me')} onChange={this.updateSetting('bio')} value={bio} type='textarea' />
        <SettingsControl
          label={t('Location')}
          onChange={this.updateSettingDirectly('location', true)}
          location={location}
          locationObject={locationObject}
          type='location'
        />
        <SettingsControl label={t('Website')} onChange={this.updateSetting('url')} value={url} />
        <SettingsControl label={t('My Skills & Interests')} renderControl={() =>
          <SkillsSection personId={currentUser.id} />} />
        <SettingsControl label={t('What I\'m learning')} renderControl={() =>
          <SkillsToLearnSection personId={currentUser.id} />} />
        <SettingsControl label={t('Contact Email')} onChange={this.updateSetting('contactEmail')} value={contactEmail} />
        <SettingsControl label={t('Contact Phone')} onChange={this.updateSetting('contactPhone')} value={contactPhone} />
        <label styleName='social-label'>{t('Social Accounts')}</label>
        <SocialControl
          label='Facebook'
          provider='facebook'
          value={facebookUrl}
          onLink={() => loginWithService('facebook')}
          updateSettingDirectly={() => this.updateSettingDirectly('facebookUrl')}
          handleUnlinkAccount={() => unlinkAccount('facebook')}
        />
        <SocialControl
          label='Twitter'
          provider='twitter'
          value={twitterName}
          updateSettingDirectly={() => this.updateSettingDirectly('twitterName')}
          handleUnlinkAccount={() => unlinkAccount('twitter')}
        />
        <SocialControl
          label='LinkedIn'
          provider='linkedin'
          value={linkedinUrl}
          updateSettingDirectly={() => this.updateSettingDirectly('linkedinUrl')}
          handleUnlinkAccount={() => unlinkAccount('linkedin')}
        />
        <div styleName='saveChanges'>
          <span styleName={changed ? 'settingChanged' : ''}>{changed ? t('Changes not saved') : t('Current settings up to date')}</span>
          <Button label={t('Save Changes')} color={changed && validateName(name) ? 'green' : 'gray'} onClick={changed && validateName(name) ? this.save : null} styleName='save-button' />
        </div>
      </div>
    )
  }
}

export default withTranslation()(EditProfileTab)
