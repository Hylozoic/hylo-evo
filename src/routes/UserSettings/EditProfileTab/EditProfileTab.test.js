/* eslint no-unused-expressions: 'off' */
import EditProfileTab, { SocialControl, linkedinPrompt, facebookPrompt } from './EditProfileTab'
import { shallow } from 'enzyme'
import React from 'react'

describe('linkedinPrompt', () => {
  var oldPrompt

  beforeAll(() => {
    oldPrompt = window.prompt
  })

  afterAll(() => {
    window.prompt = oldPrompt
  })

  it('returns a correct linkedIn Url', () => {
    const url = 'https://www.linkedin.com/in/username/'
    window.prompt = jest.fn(() => url)
    expect(linkedinPrompt()).toEqual(url)
  })

  it('rejects an incorrect linkedIn Url, calling itself again', () => {
    var count = 0
    window.prompt = jest.fn(() => {
      if (count === 0) {
        count += 1
        return 'a bad url'
      } else {
        count += 1
        return 'https://www.linkedin.com/in/username/'
      }
    })
    expect(linkedinPrompt()).toEqual('https://www.linkedin.com/in/username/')
    expect(count).toEqual(2)
    expect(window.prompt.mock.calls).toMatchSnapshot()
  })
})

describe('facebookPrompt', () => {
  var oldPrompt

  beforeAll(() => {
    oldPrompt = window.prompt
  })

  afterAll(() => {
    window.prompt = oldPrompt
  })

  it('returns a correct facebook Url', () => {
    const url = 'https://www.facebook.com/username/'
    window.prompt = jest.fn(() => url)
    expect(facebookPrompt()).toEqual(url)
  })

  it('rejects an incorrect facebook Url, calling itself again', () => {
    var count = 0
    window.prompt = jest.fn(() => {
      if (count === 0) {
        count += 1
        return 'a bad url'
      } else {
        count += 1
        return 'https://www.facebook.com/username/'
      }
    })
    expect(facebookPrompt()).toEqual('https://www.facebook.com/username/')
    expect(count).toEqual(2)
    expect(window.prompt.mock.calls).toMatchSnapshot()
  })
})

describe('EditProfileTab', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<EditProfileTab currentUser={{ name: 'Yay', locationObject: { id: 1 } }} />)
    expect(wrapper.find('Connect(UploadAttachmentButton)').length).toEqual(2)
    expect(wrapper.find('SettingsControl').length).toEqual(8)
    expect(wrapper.find('SocialControl').length).toEqual(3)
    expect(wrapper.find('Button').prop('color')).toEqual('gray')
    wrapper.setState({ changed: true })
    expect(wrapper.find('Button').prop('color')).toEqual('green')
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly without location object', () => {
    const wrapper = shallow(<EditProfileTab currentUser={{ }} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('SocialControl', () => {
  it('renders correctly without a value', () => {
    const wrapper = shallow(<SocialControl label='A Social Control' />)
    expect(wrapper.text()).toEqual('A Social ControlLink')
  })

  it('renders correctly with a value', () => {
    const wrapper = shallow(<SocialControl label='A Social Control' value='someurl.com' />)
    expect(wrapper.text()).toEqual('<Icon />A Social ControlUnlink')
  })

  it('calls handleLinkClick when link is clicked', () => {
    const wrapper = shallow(<SocialControl label='A Social Control' />)
    wrapper.instance().handleLinkClick = jest.fn()
    wrapper.find("[data-stylename='link-button']").simulate('click')
    expect(wrapper.instance().handleLinkClick).toHaveBeenCalled()
  })

  it('calls handleUnlinkClick when unlink is clicked', () => {
    const wrapper = shallow(<SocialControl label='A Social Control' value='someurl.com' />)
    wrapper.instance().handleUnlinkClick = jest.fn()
    wrapper.find("[data-stylename='link-button']").simulate('click')
    expect(wrapper.instance().handleUnlinkClick).toHaveBeenCalled()
  })

  describe('handleLinkClick', () => {
    describe('when provider is twitter', () => {
      it('and handle is entered, it updates the twitterName', () => {
        window.prompt = jest.fn(() => 'twitterhandle')
        const updateSettingDirectlyCallback = jest.fn()
        const updateSettingDirectly = jest.fn(() => updateSettingDirectlyCallback)
        const handleUnlinkAccount = jest.fn()

        const wrapper = shallow(
          <SocialControl
            label='Twitter'
            provider='twitter'
            value='twitterhandle'
            updateSettingDirectly={updateSettingDirectly}
            handleUnlinkAccount={handleUnlinkAccount}
          />
        )

        wrapper.instance().handleLinkClick()
        expect(window.prompt).toBeCalledWith('Please enter your twitter name.')
        expect(updateSettingDirectly).toHaveBeenCalled
        expect(updateSettingDirectlyCallback).toHaveBeenCalledWith('twitterhandle')
      })

      it('and handle is NOT entered, it does NOT update the twitterName', () => {
        window.prompt = jest.fn(() => false)
        const updateSettingDirectly = jest.fn(() => jest.fn())
        const handleUnlinkAccount = jest.fn()

        const wrapper = shallow(
          <SocialControl
            label='Twitter'
            provider='twitter'
            value='twitterhandle'
            updateSettingDirectly={updateSettingDirectly}
            handleUnlinkAccount={handleUnlinkAccount}
          />
        )

        wrapper.instance().handleLinkClick()
        expect(window.prompt).toBeCalledWith('Please enter your twitter name.')
        expect(updateSettingDirectly).not.toHaveBeenCalled()
      })
    })

    describe('when provider is linkedin', () => {
      it('and valid linkedinUrl is provided, it updates the linkedinUrl', () => {
        window.prompt = jest.fn(() => 'linkedin.com/test')
        const updateSettingDirectlyCallback = jest.fn()
        const updateSettingDirectly = jest.fn(() => updateSettingDirectlyCallback)
        const handleUnlinkAccount = jest.fn()

        const wrapper = shallow(
          <SocialControl
            label='LinkedIn'
            provider='linkedin'
            value='linkedin.com/test'
            updateSettingDirectly={updateSettingDirectly}
            handleUnlinkAccount={handleUnlinkAccount}
          />
        )

        wrapper.instance().handleLinkClick()
        expect(window.prompt).toBeCalledWith('Please enter the full url for your LinkedIn page.')
        expect(updateSettingDirectly).toHaveBeenCalled
        expect(updateSettingDirectlyCallback).toHaveBeenCalledWith('linkedin.com/test')
      })
    })

    describe('when provider is facebook', () => {
      it('and valid facebookUrl is provided, it updates the facebookUrl', () => {
        window.prompt = jest.fn(() => 'facebook.com/test')
        const updateSettingDirectlyCallback = jest.fn()
        const updateSettingDirectly = jest.fn(() => updateSettingDirectlyCallback)
        const handleUnlinkAccount = jest.fn()

        const wrapper = shallow(
          <SocialControl
            label='Facebook'
            provider='facebook'
            value='facebook.com/test'
            updateSettingDirectly={updateSettingDirectly}
            handleUnlinkAccount={handleUnlinkAccount}
          />
        )

        wrapper.instance().handleLinkClick()
        expect(window.prompt).toBeCalledWith('Please enter the full url for your Facebook page.')
        expect(updateSettingDirectly).toHaveBeenCalled
        expect(updateSettingDirectlyCallback).toHaveBeenCalledWith('facebook.com/test')
      })
    })
  })

  describe('handleUnlinkClick', () => {
    const handleUnlinkAccount = jest.fn()
    const updateSettingDirectlyCallback = jest.fn()
    const updateSettingDirectly = jest.fn(() => updateSettingDirectlyCallback)

    const wrapper = shallow(
      <SocialControl
        label='LinkedIn'
        provider='linkedin'
        value='linkedin.com/test'
        updateSettingDirectly={updateSettingDirectly}
        handleUnlinkAccount={handleUnlinkAccount}
      />
    )

    wrapper.instance().handleUnlinkClick()
    expect(handleUnlinkAccount).toHaveBeenCalled()
    expect(updateSettingDirectly).toHaveBeenCalled
    expect(updateSettingDirectlyCallback).toHaveBeenCalledWith(null)
  })
})
