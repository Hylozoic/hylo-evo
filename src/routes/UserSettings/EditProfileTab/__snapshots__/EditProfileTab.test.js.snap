// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfileTab renders correctly 1`] = `
<div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Your Settings
       | Hylo
    </title>
  </HelmetWrapper>
  <label
    data-stylename="label"
  >
    Your Name
  </label>
  <input
    data-stylename="name"
    onChange={[Function]}
    type="text"
    value="Yay"
  />
  <label
    data-stylename="label"
  >
    Banner and Avatar Images
  </label>
  <Connect(UploadAttachmentButton)
    data-stylename="change-banner"
    onSuccess={[Function]}
    type="userBanner"
  >
    <div
      data-stylename="banner-image"
      style={
        Object {
          "backgroundImage": "url(https://d3ngex8q79bk55.cloudfront.net/misc/default_user_banner.jpg)",
        }
      }
    >
      <Icon
        data-stylename="uploadIcon"
        name="AddImage"
      />
    </div>
  </Connect(UploadAttachmentButton)>
  <Connect(UploadAttachmentButton)
    data-stylename="change-avatar"
    onSuccess={[Function]}
    type="userAvatar"
  >
    <div
      data-stylename="avatar-image"
      style={Object {}}
    >
      <Icon
        data-stylename="uploadIcon"
        name="AddImage"
      />
    </div>
  </Connect(UploadAttachmentButton)>
  <SettingsControl
    label="Tagline"
    maxLength={60}
    onChange={[Function]}
    value=""
  />
  <SettingsControl
    label="About Me"
    onChange={[Function]}
    type="textarea"
    value=""
  />
  <SettingsControl
    label="Location"
    location=""
    locationObject={
      Object {
        "id": 1,
      }
    }
    onChange={[Function]}
    type="location"
  />
  <SettingsControl
    label="Website"
    onChange={[Function]}
    value=""
  />
  <SettingsControl
    label="My Skills & Interests"
    renderControl={[Function]}
  />
  <SettingsControl
    label="What I'm learning"
    renderControl={[Function]}
  />
  <SettingsControl
    label="Contact Email"
    onChange={[Function]}
    value=""
  />
  <SettingsControl
    label="Contact Phone"
    onChange={[Function]}
    value=""
  />
  <label
    data-stylename="social-label"
  >
    Social Accounts
  </label>
  <SocialControl
    handleUnlinkAccount={[Function]}
    label="Facebook"
    onLink={[Function]}
    provider="facebook"
    t={[Function]}
    updateSettingDirectly={[Function]}
  />
  <SocialControl
    handleUnlinkAccount={[Function]}
    label="Twitter"
    provider="twitter"
    t={[Function]}
    updateSettingDirectly={[Function]}
  />
  <SocialControl
    handleUnlinkAccount={[Function]}
    label="LinkedIn"
    provider="linkedin"
    t={[Function]}
    updateSettingDirectly={[Function]}
  />
  <div
    style={
      Object {
        "height": "80px",
      }
    }
  />
  <div
    data-stylename="saveChanges"
  >
    <span
      data-stylename="settingChanged"
    >
      Changes not saved
    </span>
    <Button
      color="green"
      data-stylename="save-button"
      label="Save Changes"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`EditProfileTab renders correctly without location object 1`] = `
<div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Your Settings
       | Hylo
    </title>
  </HelmetWrapper>
  <label
    data-stylename="label"
  >
    Your Name
  </label>
  <div
    data-stylename="name-validation"
  >
    Name must not be blank
  </div>
  <input
    data-stylename="name"
    onChange={[Function]}
    type="text"
    value=""
  />
  <label
    data-stylename="label"
  >
    Banner and Avatar Images
  </label>
  <Connect(UploadAttachmentButton)
    data-stylename="change-banner"
    onSuccess={[Function]}
    type="userBanner"
  >
    <div
      data-stylename="banner-image"
      style={
        Object {
          "backgroundImage": "url(https://d3ngex8q79bk55.cloudfront.net/misc/default_user_banner.jpg)",
        }
      }
    >
      <Icon
        data-stylename="uploadIcon"
        name="AddImage"
      />
    </div>
  </Connect(UploadAttachmentButton)>
  <Connect(UploadAttachmentButton)
    data-stylename="change-avatar"
    onSuccess={[Function]}
    type="userAvatar"
  >
    <div
      data-stylename="avatar-image"
      style={Object {}}
    >
      <Icon
        data-stylename="uploadIcon"
        name="AddImage"
      />
    </div>
  </Connect(UploadAttachmentButton)>
  <SettingsControl
    label="Tagline"
    maxLength={60}
    onChange={[Function]}
    value=""
  />
  <SettingsControl
    label="About Me"
    onChange={[Function]}
    type="textarea"
    value=""
  />
  <SettingsControl
    label="Location"
    location=""
    onChange={[Function]}
    type="location"
  />
  <SettingsControl
    label="Website"
    onChange={[Function]}
    value=""
  />
  <SettingsControl
    label="My Skills & Interests"
    renderControl={[Function]}
  />
  <SettingsControl
    label="What I'm learning"
    renderControl={[Function]}
  />
  <SettingsControl
    label="Contact Email"
    onChange={[Function]}
    value=""
  />
  <SettingsControl
    label="Contact Phone"
    onChange={[Function]}
    value=""
  />
  <label
    data-stylename="social-label"
  >
    Social Accounts
  </label>
  <SocialControl
    handleUnlinkAccount={[Function]}
    label="Facebook"
    onLink={[Function]}
    provider="facebook"
    t={[Function]}
    updateSettingDirectly={[Function]}
  />
  <SocialControl
    handleUnlinkAccount={[Function]}
    label="Twitter"
    provider="twitter"
    t={[Function]}
    updateSettingDirectly={[Function]}
  />
  <SocialControl
    handleUnlinkAccount={[Function]}
    label="LinkedIn"
    provider="linkedin"
    t={[Function]}
    updateSettingDirectly={[Function]}
  />
  <div
    style={
      Object {
        "height": "80px",
      }
    }
  />
  <div
    data-stylename="saveChanges"
  >
    <span
      data-stylename=""
    >
      Current settings up to date
    </span>
    <Button
      color="gray"
      data-stylename="save-button"
      label="Save Changes"
      onClick={null}
    />
  </div>
</div>
`;

exports[`prompt rejects an incorrect facebook Url, calling itself again 1`] = `
Array [
  Array [
    "Please enter the full url for your {{network}} page.",
  ],
  Array [
    "Invalid url. Please enter the full url for your {{network}} page.",
  ],
]
`;

exports[`prompt rejects an incorrect linkedIn Url, calling itself again 1`] = `
Array [
  Array [
    "Please enter the full url for your {{network}} page.",
  ],
  Array [
    "Invalid url. Please enter the full url for your {{network}} page.",
  ],
]
`;
